proof.dfy

```diff
@@ -148,6 +211,138 @@ module Proof {
     }
   }
 
+  predicate ReplicaInternalMsg(msg:Messages.Message)
+  {
+    || msg.PrePrepare?
+    || msg.Prepare?
+    || msg.Commit?
+  }
+
+  lemma SendersAreAlwaysFewerThanMessages(msgs:set<Messages.Message>)
+    ensures |sendersOf(msgs)| <= |msgs|
+  {
+    if |msgs| > 0 {
+      var sender :| sender in sendersOf(msgs);
+      var msgsFromSender := set msg | && msg in msgs 
+                                      && msg.sender == sender;
+      var subMsgs := msgs - msgsFromSender;
+      SendersAreAlwaysFewerThanMessages(subMsgs);
+      Library.SubsetCardinality(sendersOf(subMsgs), sendersOf(msgs) - {sender});
+      assert sendersOf(msgs) - {sender} == sendersOf(subMsgs); // Trigger
+    }
+  }
+
+  lemma WlogCommitAgreement(c: Constants, v:Variables, v':Variables, step:Step, h_step:Replica.Step,
+                            old_msg:Messages.Message, new_msg:Messages.Message)
+    requires NextStep(c, v, v', step)
+    requires c.IsReplica(step.id)
+    requires  var h_c := c.hosts[step.id].replicaConstants;
+              var h_v := v.hosts[step.id].replicaVariables;
+              var h_v' := v'.hosts[step.id].replicaVariables;
+              && Replica.NextStep(h_c, h_v, h_v', step.msgOps, h_step)
+              && h_step.SendCommitStep?
+              && Replica.SendCommit(h_c, h_v, h_v', step.msgOps, h_step.seqID)
+    requires old_msg in v.network.sentMsgs && old_msg in v'.network.sentMsgs && ReplicaInternalMsg(old_msg)
+    requires new_msg !in v.network.sentMsgs && new_msg in v'.network.sentMsgs && ReplicaInternalMsg(new_msg)
+    requires && old_msg.seqID == new_msg.seqID
+             && old_msg.view == new_msg.view
+             && old_msg.sender == new_msg.sender
+             && old_msg.Commit?
+             && new_msg.Commit?
```

Comment:
ReplicaInternalMsg can be replaced directly by Commit? since the lemma requires commit messages only.

Reply:
You are right, I will fixed this. The checks for .Commit? will be moved accordingly. Thanks :)